### PR TITLE
Feature/52 caching

### PR DESCRIPTION
**Pull request by sdorra**

### CODE REVIEW

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/PullChangesStrategy.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.cloudogu.smeagol.wiki.infrastructure;
+
+import com.cloudogu.smeagol.wiki.domain.WikiId;
+
+/**
+ * A {@link PullChangesStrategy} decides, if changes should be pulled from the remote repository of the wiki.
+ */
+public interface PullChangesStrategy {
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/PullChangesStrategy.java]: MAJOR Annotate the "PullChangesStrategy" interface with the @FunctionalInterface annotation [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS1609)

File: src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+public class TimeBasedPullChangesStrategyTest {
+
+    @Test
+    public void shouldPull() throws InterruptedException {
+        TimeBasedPullChangesStrategy strategy = new TimeBasedPullChangesStrategy(100L);
+        assertThat(strategy.shouldPull(WIKI_ID_42)).isTrue();
+        assertThat(strategy.shouldPull(WIKI_ID_42)).isFalse();
+        assertThat(strategy.shouldPull(new WikiId("42", "develop"))).isTrue();
+        Thread.sleep(101L);
```
Comment (cesmarvin) [inline src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java]: MAJOR Remove this use of "Thread.sleep()". [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS2925)

**Comment by cesmarvin** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.junit.Test;
+
+import static com.cloudogu.smeagol.wiki.DomainTestData.WIKI_ID_42;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TimeBasedPullChangesStrategyTest {
+
+    @Test
+    public void shouldPull() throws InterruptedException {
+        TimeBasedPullChangesStrategy strategy = new TimeBasedPullChangesStrategy(100L);
```
Comment (mbehlendorf) [inline src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java]: Why wait so long? Shouldn`t 1ms be enough?

File: src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.cloudogu.smeagol.wiki.infrastructure;
+
+import com.cloudogu.smeagol.wiki.domain.WikiId;
+import org.junit.Test;
+
+import static com.cloudogu.smeagol.wiki.DomainTestData.WIKI_ID_42;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TimeBasedPullChangesStrategyTest {
+
```
Comment (mbehlendorf) [inline src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java]: Maybe we should add a test: shouldPullNot ?

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategy.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    private final long duration;
+
+    public TimeBasedPullChangesStrategy(long duration) {
+        this.duration = duration;
+    }
+
+    @Override
+    public boolean shouldPull(WikiId wikiId) {
+        long timestamp = timestamps.getOrDefault(wikiId, ZERO) + duration;
```
Comment (mbehlendorf) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategy.java]: timestamp could be more meaningful like `expirationTime`

File: src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java (hunk 4 above)
Comment (sdorra) [reply src/test/java/com/cloudogu/smeagol/wiki/infrastructure/TimeBasedPullChangesStrategyTest.java]: The test for should pull contains both shouldPull returning true and false.

**Pull request merged**